src/test/scala/mesosphere/marathon/integration/setup/ServiceMock.scala

```diff
@@ -0,0 +1,199 @@
+package mesosphere.marathon.integration.setup
+
+import java.util.Date
+import javax.servlet.http.{ HttpServletRequest, HttpServletResponse }
+
+import mesosphere.marathon.integration.setup.ServiceMock.Plan
+import org.eclipse.jetty.server.handler.AbstractHandler
+import org.eclipse.jetty.server.{ Request, Server }
+import play.api.libs.json.{ Json, Writes }
+import scala.util.parsing.combinator.RegexParsers
+
+/**
+  * ServiceMock that speaks the service upgrade protocol v1 + admin handlers.
+  *
+  * GET  /v1/plan -> returns the plan
+  * PUT  /v1/plan -> continues the plan with the command flag (decision point)
+  *
+  * POST /admin/rollback -> rollback the plan
+  * PUT  /admin/next -> proceed with next step (no decision point)
+  * POST /admin/error -> add an error to the plan
+  *
+  * @param plan the plan to execute on
+  */
+class ServiceMock(plan: Plan) extends AbstractHandler {
+
+  def start(port: Int) {
+    val server = new Server(port)
+    server.setHandler(this)
+    server.start()
+    server.join()
+  }
+
+  override def handle(target: String,
+                      baseRequest: Request,
+                      request: HttpServletRequest,
+                      response: HttpServletResponse): Unit = {
+    def status[T](t: T, code: Int)(implicit writes: Writes[T]): Unit = {
+      response.getWriter.write(Json.prettyPrint(Json.toJson(t)))
+      response.setHeader("Content-Type", "application/json")
+      response.setStatus(code)
+    }
+    def ok[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 200)
+    def error[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 503)
+
+    (request.getMethod, request.getPathInfo) match {
+      case ("GET", "/v1/plan") =>
+        if (plan.isDone) ok(plan) else error(plan)
+      case ("PUT", "/v1/plan") =>
+        val cmd = Option(request.getParameter("cmd")).getOrElse("undefined")
+        cmd match {
+          case "continue" => plan.next(withContinue = true)
+          case "finalize" => plan.doFinalize()
+        }
+        ok(Json.obj("Result" -> s"Received cmd: $cmd"))
+      case ("POST", "/admin/rollback") =>
+        plan.rollback()
+        ok(Json.obj("Result" -> "Rollback"))
+      case ("PUT", "/admin/next") =>
+        plan.next(withContinue = false)
+        ok(Json.obj("Result" -> plan.status))
+      case ("POST", "/admin/error") =>
+        plan.errors ::= s"Got an error at ${new Date}"
+        ok(plan.errors)
+    }
+    baseRequest.setHandled(true)
+  }
+}
+
+object ServiceMock {
+
+  case class Plan(phases: List[Phase]) {
+    var errors: List[String] = List.empty
+    def isDone = phases.forall(_.isDone)
+    def status = currentPhase.fold("Complete")(_.status)
+    def currentPhase: Option[Phase] = phases.find(_.currentBlock.isDefined)
+    def next(withContinue: Boolean): Boolean = {
```

Comment:
Shows if the item (plan in this case) is finished.
I use this in Phase and Block and wanted to make it consistent for Plan as well.
Added scaladoc to address the gap.